### PR TITLE
Add mobile breakpoint config to Cassiopeia template and fix padding issues

### PR DESCRIPTION
**Pull request by RickR2H**
Pull Request for Issue # .

### Summary of Changes
This PR addresses two issues:

The Cassiopeia template has a left and a right position. At the moment the breakpoint at which the left and right position will stack is way to small and not functional. This is the same with modules in top-a, top-b, bottom-a and bottom-b. With this PR I made the mobile breakpoint configurable in the template settings.

template-parameters

Modules is top-a, top-b, bottom-a and bottom-b don't align with the sidebar-left and sidebar-right position of the Cassiopeia template. This PR fixes this.

align-modules

The extra mobile classes in the body, adds also a lot of extra flexibility in styling a custom template.

### Testing Instructions
1. Apply the patch and compile the CSS
2. Add multiple modules to the sidebar-left, sidebar-right, top-a, top-b, bottom-a and bottom-b
3. See if the modules in top-a, top-b, bottom-a and bottom-b align with the  sidebar-left and sidebar-right
4. Set the mobile breakpoint in te template an see if the modules respond to the set breakpoint

### Actual result BEFORE applying this Pull Request
See description above

### Expected result AFTER applying this Pull Request
See description above

### Documentation Changes Required
No

### CODE REVIEW

File: templates/cassiopeia/index.php [hunk 1]
```diff
@@ -114,6 +118,8 @@
 	. ($task ? ' task-' . $task : ' no-task')
 	. ($itemid ? ' itemid-' . $itemid : '')
 	. ($pageclass ? ' ' . $pageclass : '')
+    . ($breakPointTemplate ? ' break-tmpl-' . $breakPointTemplate : '')
+    . ($breakPointGrid ? ' break-grid-' . $breakPointGrid : '')
```
Comment (richard67) [inline templates/cassiopeia/index.php]:
```suggestion
	. ($breakPointTemplate ? ' break-tmpl-' . $breakPointTemplate : '')
	. ($breakPointGrid ? ' break-grid-' . $breakPointGrid : '')
```
Please use tabs to indent PHP code.

File: templates/cassiopeia/templateDetails.xml [hunk 2]
```diff
@@ -129,6 +129,34 @@
 					<option value="1">TPL_CASSIOPEIA_FLUID</option>
 				</field>
 
+                <field
```
Comment (richard67) [inline templates/cassiopeia/templateDetails.xml]:
```suggestion
				<field
```
Please correctly indent XML code.

File: templates/cassiopeia/templateDetails.xml [hunk 3]
```diff
@@ -129,6 +129,34 @@
...
+					description="TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION"
+					>
+					<option value="xxl">TPL_CASSIOPEIA_XXL</option>
+					<option value="xl">TPL_CASSIOPEIA_XL</option>
+					<option value="lg">TPL_CASSIOPEIA_LG</option>
+					<option value="md">TPL_CASSIOPEIA_MD</option>
+					<option value="sm">TPL_CASSIOPEIA_SM</option>
+				</field>
+
+                <field
```
Comment (richard67) [inline templates/cassiopeia/templateDetails.xml]:
```suggestion
				<field
```

Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml]: Fix in my next commit

Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml]: Fix in my next commit

File: templates/cassiopeia/index.php (hunk 1 above)
Comment (RickR2H) [reply templates/cassiopeia/index.php]: Fix in my next commit

File: templates/cassiopeia/templateDetails.xml [hunk 4]
```diff
@@ -129,6 +129,34 @@
...
+					<option value="md">TPL_CASSIOPEIA_MD</option>
+					<option value="sm">TPL_CASSIOPEIA_SM</option>
+				</field>
+
+				<field
+					name="breakPointGrid"
+					type="list"
+					default="lg"
+					label="TPL_CASSIOPEIA_BP_GRID_LABEL"
+					description="TPL_CASSIOPEIA_BP_GRID_DESCRIPTION"
```
Comment (Quy) [inline templates/cassiopeia/templateDetails.xml]:
```suggestion
					label="TPL_CASSIOPEIA_BP_TEMPLATE_LABEL"
					description="TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION"
					default="lg"
					>
					<option value="xxl">TPL_CASSIOPEIA_XXL</option>
					<option value="xl">TPL_CASSIOPEIA_XL</option>
					<option value="lg">TPL_CASSIOPEIA_LG</option>
					<option value="md">TPL_CASSIOPEIA_MD</option>
					<option value="sm">TPL_CASSIOPEIA_SM</option>
				</field>

				<field
					name="breakPointGrid"
					type="list"
					label="TPL_CASSIOPEIA_BP_GRID_LABEL"
					description="TPL_CASSIOPEIA_BP_GRID_DESCRIPTION"
					default="lg"
```

File: language/en-GB/tpl_cassiopeia.ini [hunk 5]
```diff
@@ -22,3 +22,12 @@ TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
 TPL_CASSIOPEIA_TAGLINE="Tagline"
 TPL_CASSIOPEIA_TITLE="Title (alternative to logo)"
 TPL_CASSIOPEIA_XML_DESCRIPTION="Continuing the space theme (Solarflare from 1.0, Milkyway from 1.5 and Protostar from 3.0), Cassiopeia is the Joomla 4 site template based on Bootstrap 5."
+TPL_CASSIOPEIA_BP_TEMPLATE_LABEL="Set Template Breakpoint"
```
Comment (brianteeman) [inline language/en-GB/tpl_cassiopeia.ini]: all language strings should be alpha sorted and not appended to the end of the file

File: language/en-GB/tpl_cassiopeia.ini [hunk 6]
```diff
@@ -22,3 +22,12 @@ TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
...
+TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION="Breakpoint at which the left an right position should stack above and below the content"
```
Comment (brianteeman) [inline language/en-GB/tpl_cassiopeia.ini]:
```suggestion
TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION="Breakpoint at which the left and right position should stack above or below the content."
```

File: language/en-GB/tpl_cassiopeia.ini [hunk 7]
```diff
@@ -22,3 +22,12 @@ TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
...
+TPL_CASSIOPEIA_BP_GRID_LABEL="Set Grid Breakpoint"
+TPL_CASSIOPEIA_BP_GRID_DESCRIPTION="Breakpoint at which the modules in top-a, top-b, bottom-a, and bottom-b should stack"
```
Comment (brianteeman) [inline language/en-GB/tpl_cassiopeia.ini]:
```suggestion
TPL_CASSIOPEIA_BP_GRID_DESCRIPTION="Breakpoint at which the modules in top-a, top-b, bottom-a, and bottom-b should stack."
```

File: templates/cassiopeia/templateDetails.xml (hunk 4 above)
Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml]: If I get you right, default should be last in the field parameters? Is there standard? I noticed other parameters in the file are also not in this order. Should these be changed too?

Comment (brianteeman) [reply templates/cassiopeia/templateDetails.xml]: The only thing documented is that it should start in this order

- name
- type
- label
- description

File: language/en-GB/tpl_cassiopeia.ini (hunk 5 above)
Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini]: Thanks Brian! Changed the order in my commit.

File: language/en-GB/tpl_cassiopeia.ini (hunk 7 above)
Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini]: Thanks Brian. Changes added.

File: templates/cassiopeia/templateDetails.xml (hunk 4 above)
Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml]: Clear! Thanks!

File: language/en-GB/tpl_cassiopeia.ini [hunk 8]
```diff
@@ -7,6 +7,15 @@ CASSIOPEIA="Cassiopeia Site template"
 TPL_CASSIOPEIA_BACKTOTOP="Back to Top"
 TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Link"
 TPL_CASSIOPEIA_BRAND_LABEL="Brand"
+TPL_CASSIOPEIA_BP_GRID_LABEL="Set Grid Breakpoint"
```
Comment (sandewt) [inline language/en-GB/tpl_cassiopeia.ini]: Line 15 and 16 should be alphabetical, must be:
```
TPL_CASSIOPEIA_BP_GRID_DESCRIPTION="Breakpoint at which the modules in top-a, top-b, bottom-a, or bottom-b should stack."
TPL_CASSIOPEIA_BP_GRID_LABEL="Set Grid Breakpoint"
```

Likewise line 15 and 16

And other lines...

Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini]: You are right! Missed that one!

**Comment by RickR2H** [comment]
@sandewt please test this PR 😉

File: language/en-GB/tpl_cassiopeia.ini (hunk 8 above)
Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini]: In my head description has to come after a label. Technically you ar right.

Comment (brianteeman) [reply language/en-GB/tpl_cassiopeia.ini]: catches me everytime

**Comment by ricardo1709** [comment]
I have tested this item :white_check_mark: successfully on 2575a8e6cfd49436d092459857c2802e1cb8eb16<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

File: templates/cassiopeia/templateDetails.xml [hunk 9]
```diff
@@ -129,6 +129,34 @@
 					<option value="1">TPL_CASSIOPEIA_FLUID</option>
 				</field>
 
+				<field
+					name="breakPointTemplate"
+					type="list"
+					label="TPL_CASSIOPEIA_BP_TEMPLATE_LABEL"
+					description="TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION"
+					default="lg"
```
Comment (sandewt) [inline templates/cassiopeia/templateDetails.xml:138]: Why not add `validate="options"` here, like elsewhere in Joomla ?

File: language/en-GB/tpl_cassiopeia.ini [hunk 10]
```diff
@@ -6,6 +6,15 @@
 CASSIOPEIA="Cassiopeia Site template"
 TPL_CASSIOPEIA_BACKTOTOP="Back to Top"
 TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Link"
+TPL_CASSIOPEIA_BP_GRID_DESCRIPTION="Breakpoint at which the modules in top-a, top-b, bottom-a, or bottom-b should stack."
+TPL_CASSIOPEIA_BP_GRID_LABEL="Set Grid Breakpoint"
+TPL_CASSIOPEIA_BP_LG="Large (992px)"
+TPL_CASSIOPEIA_BP_MD="Medium (768px)"
+TPL_CASSIOPEIA_BP_SM="Small (576px)"
+TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION="Breakpoint at which the left an right position should stack above and below the content."
```
Comment (sandewt) [inline language/en-GB/tpl_cassiopeia.ini]: Line 14: Should be `left and right position`. (an -> and)

Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini]: Check!

Comment (sandewt) [reply language/en-GB/tpl_cassiopeia.ini]: I see...

File: templates/cassiopeia/templateDetails.xml (hunk 9 above)
Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml:138]: Maybe it's a bot out of scope for this PR as some other fields in templateDetails.xml are also lagging the validate="options". Also the filter="" option is missing in some cases. I can change my added fields, but maybe it's better to review all the fields. @richard67 what do you think? Or should this be added in an other PR?

Comment (richard67) [reply templates/cassiopeia/templateDetails.xml:138]: For the list fields which you are adding with this PR here you should use `validate="options"` in this PR here.

> Maybe it's a bot out of scope for this PR as some other fields in templateDetails.xml are also lagging the validate="options".

I only see one list field in that XML which has not `validate="options"`, and that's the `colorName` field for the colour theme name. Feel free to make an issue or PR for that.

> Also the filter="" option is missing in some cases.

Where are they missing in your opinion?

Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml:138]: More strict is better 😉 
I've added the validate rule to the colorName field. It's such a small change and and easy fix.
@sandewt thanks for pointing out to me.

@richard67 will the tests be reset?

Comment (richard67) [reply templates/cassiopeia/templateDetails.xml:138]: Sure it needs new tests since the changes after the last tests were not just code style.

Comment (RickR2H) [reply templates/cassiopeia/templateDetails.xml:138]: Okay! Thanks Richard!

**Comment by RickR2H** [comment]
@ricardo1709 Would you please test again!

File: templates/cassiopeia/index.php [hunk 11]
```diff
@@ -31,6 +31,10 @@
 $task     = $app->input->getCmd('task', '');
 $itemid   = $app->input->getCmd('Itemid', '');
 $sitename = htmlspecialchars($app->get('sitename'), ENT_QUOTES, 'UTF-8');
+
+$breakPointTemplate = $this->params->get('breakPointTemplate', 'lg');
```
Comment (Quy) [inline templates/cassiopeia/index.php]: `breakPoint` is one word so it should be `breakpoint`.

Comment (RickR2H) [reply templates/cassiopeia/index.php]: Check! Changed in new commit.

**Comment by RickR2H** [comment]
@Quy please test 😉

**Comment by ricardo1709** [comment]
I have tested this item :white_check_mark: successfully on cbbab81b803c0b771bf166c0bffa344987bf68e1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

File: language/en-GB/tpl_cassiopeia.ini [hunk 12]
```diff
@@ -6,6 +6,15 @@
 CASSIOPEIA="Cassiopeia Site template"
 TPL_CASSIOPEIA_BACKTOTOP="Back to Top"
 TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Link"
+TPL_CASSIOPEIA_BP_GRID_DESCRIPTION="Breakpoint at which the modules in top-a, top-b, bottom-a, or bottom-b should stack."
+TPL_CASSIOPEIA_BP_GRID_LABEL="Set Grid Breakpoint"
+TPL_CASSIOPEIA_BP_LG="Large (992px)"
+TPL_CASSIOPEIA_BP_MD="Medium (768px)"
+TPL_CASSIOPEIA_BP_SM="Small (576px)"
+TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION="Breakpoint at which the left and right position should stack above and below the content."
+TPL_CASSIOPEIA_BP_TEMPLATE_LABEL="Set Template Breakpoint"
```
Comment (Quy) [inline language/en-GB/tpl_cassiopeia.ini]:
```suggestion
TPL_CASSIOPEIA_BP_GRID_LABEL="Grid Breakpoint"
TPL_CASSIOPEIA_BP_LG="Large (992px)"
TPL_CASSIOPEIA_BP_MD="Medium (768px)"
TPL_CASSIOPEIA_BP_SM="Small (576px)"
TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION="Breakpoint at which the left and right position should stack above and below the content."
TPL_CASSIOPEIA_BP_TEMPLATE_LABEL="Template Breakpoint"
```

Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini]: Less is more...

**Comment by RickR2H** [comment]
@brianteeman @ricardo1709 could you please test again.
The errors in the language strings are biting me in the you know what.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

File: language/en-GB/tpl_cassiopeia.ini [hunk 13]
```diff
@@ -6,6 +6,15 @@
 CASSIOPEIA="Cassiopeia Site template"
 TPL_CASSIOPEIA_BACKTOTOP="Back to Top"
 TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Link"
+TPL_CASSIOPEIA_BP_GRID_DESCRIPTION="Breakpoint at which the modules in top-a, top-b, bottom-a, or bottom-b should stack."
+TPL_CASSIOPEIA_BP_GRID_LABEL="Grid Breakpoint"
+TPL_CASSIOPEIA_BP_LG="Large (992px)"
+TPL_CASSIOPEIA_BP_MD="Medium (768px)"
+TPL_CASSIOPEIA_BP_SM="Small (576px)"
+TPL_CASSIOPEIA_BP_TEMPLATE_DESCRIPTION="Breakpoint at which the left and right position should stack above and below the content."
```
Comment (sandewt) [inline language/en-GB/tpl_cassiopeia.ini:14]: `above and below` -> `above or below` ?
See note @brianteeman.

Comment (brianteeman) [reply language/en-GB/tpl_cassiopeia.ini:14]: it is correct as it is

Comment (RickR2H) [reply language/en-GB/tpl_cassiopeia.ini:14]: Thanks @brianteeman

**Comment by ricardo1709** [comment]
I have tested this item :white_check_mark: successfully on 5e4253424a0f902162d19f27d1e862909b641472<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

**Comment by MSnoeren** [comment]
I have tested this item :white_check_mark: successfully on 5e4253424a0f902162d19f27d1e862909b641472<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

**Comment by bembelimen** [comment]
Why do we need options for that? I mean, can't we just do it "right" without options?

**Comment by RickR2H** [comment]
@bembelimen Imho the right wat is to make the template scalable, flexible and easy to use even for non programmers. By the way , setting a breakpoint for the left and right position at 992px is not even close to right! This is my attempt to make the template as useful as possible, so people actually start using the template in production or as a solid starting point. If the right way is building form scratch, why do we even bother building a frontend template in the first place. If we want more people to actually start using Joomla, we should make things as easy as possible and make sure thinks look good!

Nevertheless, if you have a better way we can achieve a better and flexible template then I'm all ears.

**Comment by richard67** [comment]
Added the RLDQ (release lead decision queue) label. @wilsonge Please decide if that shall go into 4.0 or not before the language freeze. If not, we have to move it to 4.1.

**Comment by RickR2H** [comment]
Would be really great to get this merged before the freeze!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34687">issues.joomla.org/tracker/joomla-cms/34687</a>.</sub>

**Comment by wilsonge** [comment]
I agree with @bembelimen this doesn't really feel like the right place for a parameter for the template. If we need to tweak the stack break and alignment let's do that directly in the scss rather than adding parameters that will be very rarely modified and require a decent amount more scss maintainer overhead.

**Comment by RickR2H** [comment]
Thanks @wilsonge for you feedback. I partially agree with you. My idea was to make the template actually functional in real live scenarios out of the box. To me it feels a bit like a crippled template that could have some great potential. Imho we now scare new users away by telling them they have to tweak the CSS first to get a decent and practical template.

**Comment by bembelimen** [comment]
I still think, it would be a valid PR which just increases the break point once if the current step is too small

**Comment by RickR2H** [comment]
> I still think, it would be a valid PR which just increases the break point once if the current step is too small

That would be a solution to make the template at least a bit more practical. What breakpoint would in your opinion be the best option, lg (992px) or xl (1200px)?

**Pull request closed without merging**